bio/dada2/filter-trim/wrapper.R

```diff
@@ -0,0 +1,76 @@
+# __author__ = "Charlie Pauvert"
+# __copyright__ = "Copyright 2020, Charlie Pauvert"
+# __email__ = "cpauvert@protonmail.com"
+# __license__ = "MIT"
+
+# Snakemake wrapper for filtering single or paired-end reads using dada2 filterAndTrim function.
+
+# Sink the stderr and stdout to the snakemake log file
+# https://stackoverflow.com/a/48173272
+log.file<-file(snakemake@log[[1]],open="wt")
+sink(log.file)
+sink(log.file,type="message")
+
+library(dada2)
+
+# Custom function to check if homogeneous compression suffixes are given
+all_compressed<-function(f){
+	if(!is.null(f)){
+		compressed<-grepl(".gz$|.bz2$", f)
+		if(all(compressed)){
+			return(TRUE)
+		} else if(all(!compressed)){
+			return(FALSE)
+		} else {
+			stop("Non-homogeneous suffixes (.gz or .bz2) for some files.")
+		}
+
+	} else {
+		return(NULL)
+	}
+}
+
+# Prepare arguments (no matter the order)
+args<-list(
+        fwd = snakemake@input[["fwd"]],
+        filt = snakemake@output[["filt"]],
+        multithread=snakemake@threads
+)
+# Test if paired end input is passed
+if(!is.null(snakemake@input[["rev"]]) & !is.null(snakemake@output[["filt_rev"]])){
+        args<-c(args,
+            rev = snakemake@input[["rev"]],
+            filt.rev = snakemake@output[["filt_rev"]]
+            )
+}
+# Check if extra params are passed
+if(length(snakemake@params) > 0 ){
+    # Keeping only the named elements of the list for do.call()
+    extra<-snakemake@params[ names(snakemake@params) != "" ]
+# Check if 'compress=' option is passed
+if(!is.null(extra[["compress"]])){
+    stop("Remove the `compress` option from the Snakefile.\n",
+    "The `compress` option is implicitly set here from the file extension.")
+    } else {
+    # Check if output files are given as compressed files
+    # ex: in se version, all(TRUE, NULL) gives TRUE
+    extra[["compress"]]<-all(
+    			all_compressed(args[["filt"]]),
+    			all_compressed(args[["filt.rev"]])
+    			)
+    }
```

Comment:
I just realized, that it is actually OK to have different `compress` status for the two paired-end output files, according to the docs. I don't know why a user would want that, but no need to force the user to avoid something that the software permits. So the following should do the trick:
```suggestion
    extra[["compress"]]<-c(
        endsWith(args[["filt"]], '.gz'),
        endsWith(args[["filt.rev"]], 'gz')
    )
```

The only case I am not sure about is the single-end case, because then `compress` is a two-entry vector. This could be easier, if we remove the explicit naming of inputs and simply check whether we have one or two inputs to decide whether it is single end or paired end input. So something like:
```
input:
    "{sample}.1.fastq.gz",
    "{sample}.1.fastq.gz"
output:
    filtered=["filtered/{sample}.1.fastq.gz", "filtered/{sample}.2.fastq.gz"],
    stats="reports/{sample}.tsv"
```
And then check the length of `snakemake@input` to decide whether only `fwd` is assigned to, or if a second entry is available for `rev`.

Reply:
In the latter case, we can simply use `extra[["compress"]]<-lapply(snakemake@output[["filtered"]], endsWith, suffix='.gz')` to get a list of appropriate length.

Reply:
Ok for your suggestion! But no need to remove the explicit naming.
Actually we can exploit the `NULL` property!
For instance: `c(TRUE, NULL)` gives `c(TRUE)` in R!
Meaning that combining this magic and yours we get something like this:
```r
extra[["compress"]]<-c(
        endsWith(args[["filt"]], '.gz'),
        if(is.null(args[["filt.rev"]])) NULL else {endsWith(args[["filt.rev"]], 'gz')}
    )
```

Reply:
Done in the last commit!